### PR TITLE
build: bump to go1.9rc1

### PR DESCRIPTION
**Pull request by petermattis**

### CODE REVIEW

File: build/Dockerfile [hunk 1]
```diff
@@ -103,16 +103,13 @@ RUN git clone --depth 1 https://github.com/tpoechtrager/osxcross.git \
...
- && echo '5f5dea2447e7dcfdc50fa6b94c512e58bfba5673c039259fd843f68829d99fa6  golang.tar.gz' | sha256sum -c - \
+RUN curl -fsSL https://storage.googleapis.com/golang/go1.9rc1.src.tar.gz -o golang.tar.gz \
+ && echo '87717598ea60cc6143afa48f141f7e1308e196b71862028e710b910f376b452e golang.tar.gz' | sha256sum -c - \
  && tar -C /usr/local -xzf golang.tar.gz \
  && rm golang.tar.gz \
  && cd /usr/local/go/src \
- && patch -p2 -i /parallelbuilds-go1.8.patch \
  && GOROOT_BOOTSTRAP=$(go env GOROOT) CC=clang CXX=clang++ ./make.bash \
- && rm -rf /*.patch \
- && apt-get autoremove -y golang
```
Comment (tamird) [inline build/Dockerfile]: why?

**Review by tamird — APPROVED** [review]

**Comment by tamird** [comment]
Also, shouldn't we do this everywhere? Bootstrap scripts and the like?

**Comment by petermattis** [comment]
I made this change in order to test go1.9rc1. Still to be determined is whether we want to merge it. I'm thumbs up on doing so. The alternative is to wait until the start of the 1.2 cycle to upgrade to go1.9. That will put us significantly behind the Go release. When 1.2 comes out, go1.10 will be released.

If we're going to using go1.9 for 1.1, we should make the switch early. Doing it now would give us 2 months of experience with go1.9. Note that Google switches all of their internal users of Go to the release candidates as they are released. So go1.9rc1 is being used in production inside Google.

File: build/Dockerfile (hunk 1 above)
Comment (petermattis) [inline build/Dockerfile]: This was from another branch where I was debugging why patches weren't applying. Reverted.

File: build/Dockerfile [hunk 2]
```diff
@@ -103,13 +103,11 @@ RUN git clone --depth 1 https://github.com/tpoechtrager/osxcross.git \
...
-RUN curl -fsSL https://storage.googleapis.com/golang/go1.8.3.src.tar.gz -o golang.tar.gz \
- && echo '5f5dea2447e7dcfdc50fa6b94c512e58bfba5673c039259fd843f68829d99fa6  golang.tar.gz' | sha256sum -c - \
+RUN curl -fsSL https://storage.googleapis.com/golang/go1.9rc1.src.tar.gz -o golang.tar.gz \
+ && echo '87717598ea60cc6143afa48f141f7e1308e196b71862028e710b910f376b452e golang.tar.gz' | sha256sum -c - \
  && tar -C /usr/local -xzf golang.tar.gz \
  && rm golang.tar.gz \
  && cd /usr/local/go/src \
- && patch -p2 -i /parallelbuilds-go1.8.patch \
  && GOROOT_BOOTSTRAP=$(go env GOROOT) CC=clang CXX=clang++ ./make.bash \
  && rm -rf /*.patch \
```
Comment (tamird) [inline build/Dockerfile]: this line can be removed, though

**Review by tamird — APPROVED** [review]
I'm :+1: on merging this as well.

File: build/Dockerfile (hunk 2 above)
Comment (petermattis) [inline build/Dockerfile]: Yeah, though we might be adding it back soon.

**Comment by benesch** [comment]
👍

**Review by a-robinson — APPROVED** [review]
LGTM, but this makes our release cycle's 2 month offset from the go release cycle seem less ideal than, say a 3 month offset, given that 1.9 won't be properly released until after our feature freeze.

**Comment by danhhz** [comment]
👍

**Comment by benesch** [comment]
The Go 1.9 release cycle is running 25 days behind schedule AIUI, though that just lends more weight to @a-robinson's argument for a three month release cycle.

**Review by rjnn — APPROVED** [review]
We should merge this so that we can get all our tests and clusters running Go 1.9 sooner rather than later. We can always pull it if there's an actual regression somewhere.

**Comment by BramGruneir** [comment]
Perhaps we should consider altering our release schedule to coincide better with golang's.
I'm 👎 unless there is a compelling reason (perf perhaps). 
The last thing we want to is have to issue an emergency 1.1.1 due to golang having to quickly cut a 1.9.1 to fix some security hole or other pressing issue.

**Comment by petermattis** [comment]
> The last thing we want to is have to issue an emergency 1.1.1 due to golang having to quickly cut a 1.9.1 to fix some security hole or other pressing issue.

My impression is that historically this hasn't been an issue. And security issues often affect multiple releases of Go. That is, if go1.9.1 is released to fix a security issue, there will likely be a go1.8.x release as well. The concern in my mind is whether go1.9 adds some sort of instability. But I'd hope we'd discover that with 2 months of internal testing, along with the betas and release candidates.

**Comment by bdarnell** [comment]
Is there a compelling reason to upgrade to go 1.9 at this point in the cycle? I'd generally prefer to upgrade at the beginning of a development cycle instead of right before feature freeze (even if that means we're not on the cutting edge). I was under the impression that our release cycle was timed to facilitate that, by putting the start of a new cycle just after Go's release. If we want to pick up new versions of Go quickly I think we should shift our release timing to allow a little more room between their release and our freeze.

**Comment by petermattis** [comment]
@bdarnell (and others in the anti-upgrade camp) How much time would you like between the release of a new Go version and our adoption?

**Comment by tschottdorf** [comment]
I'm 👍, have been using 1.9 locally for a while without any problems (blamable on 1.9), including on backup/restore testing. Beginning of feature freeze doesn't seem like a bad time to bump, too - we're doing more testing/debugging/fixing in that time period anyway. Minor plus, it would mean 1.1 gets a decent https+insecure pprof endpoint which makes it more fun to do perf work/debug slowness.

**Comment by petermattis** [comment]
> Is there a compelling reason to upgrade to go 1.9 at this point in the cycle?

In comparison to Go previous releases, the performance bump from go1.9 seems to be less compelling.

**Comment by bdarnell** [comment]
> How much time would you like between the release of a new Go version and our adoption?

I'd like for it to be *released* for one thing. Early in a cycle, adopting a Go RC could make sense, but I don't like upgrading the Go version post-freeze, and I don't like using an RC instead of a final release just to get something in before the freeze. The current alignment of the schedules allows no room for error. If Go releases on time, the new release comes out on the day of our feature freeze; any delays push it into our stabilization window. 

This upgrade also causes disruptive churn in everyone's development environments (package managers generally don't get updated until the final release so you have to install it manually, etc), and we're trying to avoid a repeat of 1.0's makefile rewrites that landed at the end of that development window.

**Comment by petermattis** [comment]
> I'd like for it to be released for one thing.

The Go release candidates are much more stable than our release candidates. It is useful to be conservative, but I wonder if this is too conservative. Your point about disruptive churn is valid. Have we ever noticed a stability or performance issue arise from a Go upgrade? I can't think of one.

**Comment by BramGruneir** [comment]
`Past performance is not an indicator of future outcomes.'

**Comment by tamird** [comment]
-Albert Einstein?

**Comment by bdarnell** [comment]
> Have we ever noticed a stability or performance issue arise from a Go upgrade? I can't think of one.

Neither can I. But even though go's release candidates are generally solid, I wouldn't feel comfortable shipping our 1.1-final built with a release candidate of Go. That means we're either going to need to upgrade again before 1.1 is finalized, or downgrade back to 1.8.3. How late in the process do we allow that decision to be made? 

If we were getting significant benefit from this upgrade then I'd probably be OK with it (in fact I'd have argued for upgrading to a beta release even earlier). But the case for upgrading urgently seems pretty weak to me.

**Comment by danhhz** [comment]
The pprof improvements might be useful for the august bug fixing, but I can always use a different go version locally. And the encoding/csv allocation improvements might help with csv import (some preliminary profiling suggests it would in the single node case)

OTOH, I don't know how convincing these are and I am quite concerned about the cost of dev environment churn

> That means we're either going to need to upgrade again before 1.1 is finalized, or downgrade back to 1.8.3. How late in the process do we allow that decision to be made?

For my money, this is a non-issue. If we're on a 1.9 release candidate and it's been stable for two months, switching to the final 1.9 release (even very very late in the process) is fine. This feels too conservative to me.

**Comment by petermattis** [comment]
Closing for now. Will re-open when the 1.2 tree opens.

**Pull request closed without merging**